deva.1
.TH DEVA 1

.SH NAME
deva \- Deva is a small, simple, interpreted, dynamic programming language.

.SH SYNOPSIS
\fBdeva\fP [\--help] [\--version | \-v] [\--no-dvc] [\--compile-only | \-c] [\--options] \fIinput-file

.SH DESCRIPTION
\fIDeva\fP is a small, simple, interpreted, dynamic programming language. It is similar to C in syntax while semantically similar to Python and other dynamic languages. It is embeddable in C++ programs or usable on its own. Deva is a multi-paradigm language, supporting procedural (imperative), object-oriented and functional language features.

The key design goals of Deva were to 1) be easily embeddable in C++, and 2) to provide a language that is as simple as possible in both implementation and use, while providing basic support for key language features.

For a full discussion of the language and its features, you are referred to 'The Deva Language v2' document for language documentation.

.SH OPTIONS
\fPDeva supports the following options:
.TP
\fB--help\fP
Print help
.TP
\fB--version, \-v\fP
Print version information
.TP
\fB-no-dvc\fP
Do not generate a .dvc byte-code compiled file
.TP
\fB--compile-only, \-c\fP
Compile only, do not execute
.TP
\fB--options\fP
Options to pass to the source program
.TP
\fIinput-file\fP
Input deva source code file

.SH ENVIRONMENT VARIABLES 
\fIDEVA\fP

If the DEVA environment variable is set, deva will look for files and modules in the paths pointed to by the environment variable immediately after looking in the working directory.

.SH COPYRIGHT
deva is Copyright (c) 2011-2011 Joshua C. Shepard

.SH LICENSE

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

.SH AUTHORS
Joshua C. Shepard

E-mail : josh.shepard@gmail.com



NEWS
release 0.9.9
NOTE: THIS IS LIKELY THE FINAL DEVA V.1 SERIES RELEASE. I have started work on deva 2, which is a (nearly) complete re-write. All the deva v1 sources will be moved into a sub-directory and the main directory will become the deva 2 tree.
Deva v1 was a mostly successful experiment, but was written in a somewhat ad-hoc manner and as I expanded my use of it as a language it became apparent that there are several areas in need of improvement: 1) the parser is even slower than expected. the old boost spirit parser simply creates too much overhead, 2)there are several language issues that I would like to fix, but are difficult with the current parser, 3) the execution engine is not a 'real' VM and the bytecode format falls somewhere between a tree-driven interpreter and 'real' bytecode (meaning that execution is SLOW as well), 4) etc etc. Deva 2 is my attempt to address these a produce a more "real-world-ready" small language.

- fixes to prevent compilation of statements that have no effect (other than possible side effects) such as 'a;' or 'a.b[0];' and changes to tests to make them work with this fix
- fixed misc bugs around dot ops, calls and key ops ('[]') on both sides of assignments
- added test 69 to exercise these fixes
- updated test 7 baseline to current state (though the updated part is still broken: 'a().b' works now, but leaves the result on the stack)
- prevent assignment into function calls (and math-op assignments e.g. +=)
- fix for some assignment sequences (e.g. 'a.b.c[0].d().e = 0')
- bug fix: map methods 'rewind' and 'next' wouldn't work on classes and instances, meaning you couldn't iterate their contents in a 'for' loop
- bug fix: new fcns/methods with the same name as existing fcns/methods _didn't_ override the existing ones. (defun was calling 'Scope::AddObject' without orphaning any existing object of the same name)
- added rudimentary 'devash' deva shell
- renamed 'import' builtin to 'importmodule' so that it doesn't conflict with the keyword (duh)
- fixed slices: enabled negative indices to work properly, added '$' as a symbol meaning 'end-of-list' (rather than empty indices, like python uses), fixed some bugs and updated test code (json.dv)
- added a slice test (test 68) which includes euler's sieve among other things
- changed exit() builtin to have a default value of 0 for the arg
- updated the language doc
- enable nested functions!! this was *almost* just a matter of letting the grammar accept them, with a couple of very minor tweaks in the compiler and execution engine to handle them correctly
- added test 67 to test nested functions and updated the language doc
- created an 'import' built-in and changed the code-gen for the 'import' statement to call it
- removed un-needed/used instructions/op-codes: break, output and import
- fixes 'continue' so that it corrects for scopes it jumps out of
- changed the code-gen for 'break' statements to generate 'leave' ops and a 'jmp' (like continue) and deprecated the 'break' instruction (to be removed completely before a 1.0 release)
- added test 66 to test 'continue' more thoroughly (and 'break' too)
- fixed bug in range() built-in: range generated was [start, ..., end] when it should have been [start, ... end-1]
- added Executor::PrintDataStack() method for use in devadb stack printing and viewing the deva stack from gdb
- fixed serious problems with 'break'. this involved making sure that scopes are created for all flow control and loop construct bodies and fixing 'for' loop code-gen to work with breaks. (i would like to move more of the break handling from the execution engine into code-gen, but these fixes enable break to work properly).
- fixed runtests.dv and glob.dv which were improperly defining variables inside 'if' and 'else' bodies and then using them outside the bodies (which worked before, though it shouldn't have, because no scope was generated for single-line if/else bodies)
- added test 65 which more thoroughly tests break statements
- changed test harness to be able to run valgrind against deva in execution-only or compile-and-execute modes. (compile-and-execute takes a LONG time to complete all tests, so you only want to run it when necessary)
- made two fixes to make compilation valgrind clean
- removed all valgrind suppressions, which aren't needed with gcc 4.4.3 and valgrind 3.6.0 (and boost 1.4.1)
- fixed bug in _re module where trying to compile an invalid regular expression resulted in an (un-caught) boost exception instead of a deva error
- added support for breakpoints using short filenames. for guaranteed match a full path would still be preferred (for all except the main/entry file, which is always a short-name)
- fixed the devadb 'print' command to print strings inside single quotes
- rewrote string.split properly and added a bunch of test cases for it (to test 33)
- also fixed utility fcn split_env_var_paths to use the same algorithm
- changed printing output for vectors and maps (incl classes and instances) to print string values inside single quotes. improves readability
- updated many test baselines for this new output
- fixed files that had mixed tabs & spaces for indentation
- fixed bug in string.split() which would result in bogus items
- added support for special methods 'str' and 'repr' on user-defined classes (which are called by built-ins str() and print(), respectively)
- added test for 'str' and 'repr' special methods
- updated language doc
- fixed breakpoint command in devadb to work if only a line number is given (sets bp in current file at that line)
- added editline (BSD licensed readline) support to devadb
- added readline support to 'todo.py'
- moved ICE when all scopes are not clear at ScopeTable deletion into a debug only block
- fixed the baseline for test 31
- added factorial.dv illustrating recursive, iterative and 'lispish' (using reduce()) factorial implementations in deva

src/deva.cpp
// Copyright (c) 2010 Joshua C. Shepard
// 
// Permission is hereby granted, free of charge, to any person
// obtaining a copy of this software and associated documentation
// files (the "Software"), to deal in the Software without
// restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following
// conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
// OTHER DEALINGS IN THE SOFTWARE.

// deva.cpp
// deva language executor, main program
// created by jcs, december 09 , 2010 

// TODO: 
// * quick exit, don't bother to clean up, OS will reclaim memory???
//

#include "api.h"
#include "module_os.h"
#include "module_bit.h"
#include "module_math.h"
#include "module_re.h"

#include <iostream>
#include <vector>
#include <boost/format.hpp>
#include <boost/program_options/options_description.hpp>
#include <boost/program_options/positional_options.hpp>
#include <boost/program_options/variables_map.hpp>
#include <boost/program_options/parsers.hpp>

using namespace std;
namespace po = boost::program_options;

using namespace deva;
using namespace deva_compile;

#ifdef MS_WINDOWS
#define chdir(x) _chdir(x)
#include <direct.h>
#endif

/////////////////////////////////////////////////////////////////////////////
// globals
/////////////////////////////////////////////////////////////////////////////
// global object to track current filename
const char* deva::current_file;

namespace deva
{
int _argc;
char** _argv;
}

// global to enable tracing refcounts
bool deva::reftrace;

/////////////////////////////////////////////////////////////////////////////
// functions
/////////////////////////////////////////////////////////////////////////////

// Main entry point for this example
//
int ANTLR3_CDECL main( int argc, char *argv[] )
{
	deva::_argc = argc;
	deva::_argv = argv;

	// declare the command line options
	bool debug_dump = false;
	bool trace = false;
	reftrace = false;
	bool show_ast = false;
	bool no_dvc = false;
	bool disasm = false;
	bool compile_only = false;
	string output;
	string input;
	vector<string> inputs;
	po::options_description desc( "Supported options" );
	desc.add_options()
		( "help", "help message" )
		( "version,v", "display program version" )
		( "no-dvc", "do NOT write a .dvc compiled byte-code file to disk" )
		( "compile-only,c", "compile only, do not execute" )
		( "disasm", "disassemble" )
#ifdef DEBUG
		( "trace", "show execution trace" )
		( "reftrace", "show refcount trace" )
		( "show-ast", "show the AST" )
		( "debug-dump", "turn internal debug output on" )
#endif
		( "input", po::value<string>( &input ), "input filename" )
		( "options", po::value<vector<string> >( &inputs )->composing(), "options to pass to the deva program" )
		;
	po::positional_options_description p;
	p.add( "input", 1 ).add( "options", -1 );
	po::variables_map vm;
	try
	{
		po::parsed_options parsed = po::command_line_parser( argc, argv ).options( desc ).allow_unregistered().positional( p ).run();
		po::store( parsed, vm );

	}
	catch( po::error & e )
	{
		cout << e.what() << endl;
		return 1;
	}
	po::notify( vm );

	// handle the command line args
	if( vm.count( "help" ) )
	{
		cout << desc << endl;
		return 1;
	}
	if( vm.count( "version" ) )
	{
		// dump the version number
		cout << "deva " << DEVA_VERSION << endl;
		return 1;
	}
	if( vm.count( "no-dvc" ) )
	{
		no_dvc = true;
	}
	if( vm.count( "disasm" ) )
	{
		disasm = true;
	}
#ifdef DEBUG
	if( vm.count( "trace" ) )
	{
		trace = true;
	}
	if( vm.count( "reftrace" ) )
	{
		reftrace = true;
	}
	if( vm.count( "show-ast" ) )
	{
		show_ast = true;
	}
	if( vm.count( "debug-dump" ) )
	{
		debug_dump = true;
	}
#endif
	if( vm.count( "compile-only" ) )
	{
		compile_only = true;
	}
	// must be an input file specified
	if( !vm.count( "input" ) )
	{
		cout << "usage: deva [options] <input_file>" << endl;
		cout << "(use option --help for more information)" << endl;
		return 1;
	}

	// get the filename part and the directory part
	string in_dir = get_dir_part( input );
	string fname = get_file_part( input );
	string ext = get_extension( fname );

	// change cwd to the directory
	// if the input wasn't a full path:
	if( in_dir[0] != '/' )
	{
		string cwd = get_cwd();
		string dir = join_paths( cwd, in_dir );
		if( chdir( dir.c_str() ) != 0 ) 
		{
			cout << "error: unable to change the current working directory to " << dir.c_str() << endl;
			return -1;
		}
	}
	else
	{
		if( chdir( in_dir.c_str() ) != 0 )
		{
			cout << "error: unable to change the current working directory to " << in_dir.c_str() << endl;
			return -1;
		}
	}

//	bool use_dvc = false;
//	unsigned char* code;
//	if( ext != ".dvc" )
//	{
//		// TODO: compile?
//	}

	ParseReturnValue prv;
	PassOneReturnValue p1rv;
	try
	{
		// parse the file
		prv = Parse( fname.c_str() );

		if( prv.successful )
		{
			PassOneFlags p1f; // currently no pass one flags
			PassTwoFlags p2f;
			p2f.trace = trace;

			// PASS ONE: build the symbol table and check semantics
			p1rv = PassOne( prv, p1f );

			// diagnostic information output:
			// print the text repr of the tree?
			if( show_ast )
			{
				cout << "AST:" << endl;
				pANTLR3_STRING s = p1rv.nodes->root->toStringTree( p1rv.nodes->root );
				ANTLR3_FPRINTF( stdout, "%s\n", (char*)s->chars );
			}

			// PASS TWO: compile
			ex = new Executor();
			Code* code = PassTwo( "", p1rv, p2f );

			// debug dumps
#ifdef DEBUG
			if( debug_dump /*&& verbosity == 3*/ )
			{
				// dump the symbol tables...
				semantics->DumpSymbolTable();
				// dump the constant data pool
				ex->DumpConstantPool();
				// dump the function objects
				ex->DumpFunctions();
			}
#endif

			if( disasm )
			{
				ex->Decode( code );
			}

			// free compile-time objects before executing code
			// free parser, compile memory
			FreeParseReturnValue( prv );
			FreePassOneReturnValue( p1rv );
			delete compiler;
			compiler = NULL;
			delete semantics;
			semantics = NULL;

			// execute the code
			if( !compile_only )
			{
				ex->AddNativeModule( "os", GetModuleOsFunction );
				ex->AddNativeModule( "bit", GetModuleBitFunction );
				ex->AddNativeModule( "math", GetModuleMathFunction );
				ex->AddNativeModule( "_re", GetModuleReFunction );
				ex->Execute( code );
			}
			else
				delete code;
		}
	}
	catch( SemanticException & e )
	{
		// display an error
		emit_error( e );

		// free parser, compiler memory
		FreeParseReturnValue( prv );
		FreePassOneReturnValue( p1rv );
		// free the semantics objects (symbol table et al), compiler, executor
		delete ex;
		delete compiler;
		delete semantics;

		exit( -1 );
	}
	catch( ICE & e )
	{
		// display an error
		emit_error( e );

		// free parser, compile memory
		FreeParseReturnValue( prv );
		FreePassOneReturnValue( p1rv );
		// free the semantics objects (symbol table et al), compiler, executor
		delete ex;
		delete compiler;
		delete semantics;

		exit( -1 );
	}
	catch( RuntimeException & e )
	{
		// display an error
		emit_error( e );

		// dump the stack trace
		if( ex )
			ex->DumpTrace( cerr );

		// free parser, compile memory
		FreeParseReturnValue( prv );
		FreePassOneReturnValue( p1rv );
		// free the semantics objects (symbol table et al), compiler, executor
		delete ex;
		delete compiler;
		delete semantics;

		exit( -1 );
	}

	// free the compiler
	delete ex;

	return 0;
}
